File: 0456-132 Pattern/0456-132 Pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        int n = nums.size();
        if (n < 3)
        {
            return false;
        }
        
        stack<int> St;
        int s2 = INT_MIN;
        for (int i = n - 1; i >= 0; --i)
        {
            if (nums[i] < s2)
            {
                return true;
            }
            
            while (!St.empty() && St.top() < nums[i])
            {
                s2 = St.top();
                St.pop();
            }
            
            St.push(nums[i]);
        }
        
        return false;
    }
};

File: 0401-0500/0456-132 Pattern/0456-132 Pattern.py
import math
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        St = []
        s2 = -math.inf
        for num in nums[::-1]:
            if num < s2:
                return True
            while St and St[-1] < num:
                s2 = St.pop()
            St.append(num)
        return False
